
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>Genes Â· GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.2">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="panels.html" />
    
    
    <link rel="prev" href="variants.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Introduction
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../install.html">
            
                <a href="../install.html">
            
                    
                    Installation
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../faq.html">
            
                <a href="../faq.html">
            
                    
                    FAQ
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../features/">
            
                <a href="../features/">
            
                    
                    Features
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../features/hpo.html">
            
                <a href="../features/hpo.html">
            
                    
                    HPO phenotypes
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="./">
            
                <a href="./">
            
                    
                    User Guide
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" data-path="getting-started.html">
            
                <a href="getting-started.html">
            
                    
                    Getting Started
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="using-scout.html">
            
                <a href="using-scout.html">
            
                    
                    General Usage
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.3" data-path="pages.html">
            
                <a href="pages.html">
            
                    
                    Pages
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.4" data-path="institutes.html">
            
                <a href="institutes.html">
            
                    
                    Institutes
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.5" data-path="users.html">
            
                <a href="users.html">
            
                    
                    Users
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.6" data-path="cases.html">
            
                <a href="cases.html">
            
                    
                    Cases
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.7" data-path="variants.html">
            
                <a href="variants.html">
            
                    
                    Variants
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.5.8" data-path="genes.html">
            
                <a href="genes.html">
            
                    
                    Genes
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.9" data-path="panels.html">
            
                <a href="panels.html">
            
                    
                    Gene Panels
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.10" data-path="annotations.html">
            
                <a href="annotations.html">
            
                    
                    Annotations
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.11" data-path="acmg-criterias.html">
            
                <a href="acmg-criterias.html">
            
                    
                    ACMG criterias
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="../admin-guide/">
            
                <a href="../admin-guide/">
            
                    
                    Admin Guide
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1" data-path="../admin-guide/setup-scout.html">
            
                <a href="../admin-guide/setup-scout.html">
            
                    
                    Setup
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.2" data-path="../admin-guide/loading.html">
            
                <a href="../admin-guide/loading.html">
            
                    
                    Loading
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.3" data-path="../admin-guide/load-config.html">
            
                <a href="../admin-guide/load-config.html">
            
                    
                    Load config
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.4" data-path="../admin-guide/genes.html">
            
                <a href="../admin-guide/genes.html">
            
                    
                    Gene definitions
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.5" data-path="../admin-guide/hpo.html">
            
                <a href="../admin-guide/hpo.html">
            
                    
                    HPO terms
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.6" data-path="../admin-guide/annotations.html">
            
                <a href="../admin-guide/annotations.html">
            
                    
                    Annotations
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.7" data-path="../admin-guide/server.html">
            
                <a href="../admin-guide/server.html">
            
                    
                    Server
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.8" data-path="../admin-guide/chanjo_coverage_integration.html">
            
                <a href="../admin-guide/chanjo_coverage_integration.html">
            
                    
                    Chanjo coverage integration
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.9" data-path="../admin-guide/indexes.html">
            
                <a href="../admin-guide/indexes.html">
            
                    
                    Indexes
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.10" data-path="../admin-guide/institute.html">
            
                <a href="../admin-guide/institute.html">
            
                    
                    Institutes
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="../blog/">
            
                <a href="../blog/">
            
                    
                    Blog
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.1" data-path="../blog/new-3.0.html">
            
                <a href="../blog/new-3.0.html">
            
                    
                    What's new in 3.0?
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.2" data-path="../blog/new-3.1.html">
            
                <a href="../blog/new-3.1.html">
            
                    
                    What's new in 3.1?
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.3" data-path="../blog/new-3.2.html">
            
                <a href="../blog/new-3.2.html">
            
                    
                    What's new in 3.2?
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.4" data-path="../blog/new-3.3.html">
            
                <a href="../blog/new-3.3.html">
            
                    
                    What's new in 3.3?
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.5" data-path="../blog/new-3.4.html">
            
                <a href="../blog/new-3.4.html">
            
                    
                    What's new in 3.4?
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.6" data-path="../blog/new-3.5.html">
            
                <a href="../blog/new-3.5.html">
            
                    
                    What's new in 3.5?
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.7" data-path="../blog/new-4.0.0.html">
            
                <a href="../blog/new-4.0.0.html">
            
                    
                    What's new in 4.0?
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >Genes</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="genes-and-transcripts">Genes and transcripts</h1>
<p>Scout stores information about genes and transcripts. The information is collected from a couple of resources, these can be updated manually if desired. Definition of what genes that exists and their correct id and symbols are collected from <a href="www.genenames.org">HGNC</a>. Unfortunately HGNC does only maintain a distribution for GRCh38, at this time (mid 2018) there are many resources that lack support for build 38 so many investigators still use build 37. We therefore load two gene sets, one for each build.</p>
<h2 id="genes-with-missing-refseq-transcripts">Genes with missing RefSeq transcripts</h2>
<p>There is a issue where this problem is discussed <a href="https://github.com/Clinical-Genomics/scout/issues/570" target="_blank">here</a>, please read that since it is important.
When we first look at all genes we could see the following:</p>
<p><code>Nr of genes: 33578</code>
<code>Nr without transcripts: 11048</code></p>
<p>So 1/3 of all genes are missing any refseq transcript. This came out as a high number at first, after some thought one might realise that many of these genes are not protein coding etc.</p>
<p>If we look at disease causing genes from OMIM there are 16 genes that are missing refseq id for any transcript in ensembl build 37, 7 genes in build 38.</p>
<p>The genes in OMIM without refseq transcripts in build 37 are the following:</p>
<pre><code>TTC25
PTPRQ
SRD5A2
PIGY
FGF16
TRAC
PADI6
GDF1
TUBB3
IGHG2
IGHM
IGKC
FCGR2C
KMT2B
NEFL
NR2E3
</code></pre><p>In the gene panels there are one more gene outside OMIM wihtout refseq transcripts: <code>MAP3K14</code></p>
<p>We will here look at all these genes in detail:</p>
<h3 id="ttc25">TTC25</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> 37</td>
<td>HG185_PATCH</td>
<td>ENSG00000260703</td>
<td>TTC25</td>
<td>ENST00000569427</td>
<td>-</td>
</tr>
<tr>
<td> 37</td>
<td>HG185_PATCH</td>
<td>ENSG00000260703</td>
<td>TTC25</td>
<td>ENST00000561994</td>
<td>-</td>
</tr>
<tr>
<td> 37</td>
<td>HG185_PATCH</td>
<td>ENSG00000260703</td>
<td>TTC25</td>
<td>ENST00000569541</td>
<td>NM_031421</td>
</tr>
<tr>
<td> 37</td>
<td>HG185_PATCH</td>
<td>ENSG00000260703</td>
<td>TTC25</td>
<td>ENST00000565052</td>
<td>-</td>
</tr>
<tr>
<td> 37</td>
<td>17</td>
<td>ENSG00000204815</td>
<td>TTC25</td>
<td>ENST00000593239</td>
<td>-</td>
</tr>
<tr>
<td> 37</td>
<td>17</td>
<td>ENSG00000204815</td>
<td>TTC25</td>
<td>ENST00000591658</td>
<td>-</td>
</tr>
<tr>
<td> 37</td>
<td>17</td>
<td>ENSG00000204815</td>
<td>TTC25</td>
<td>ENST00000585530</td>
<td>-</td>
</tr>
<tr>
<td> 37</td>
<td>17</td>
<td>ENSG00000204815</td>
<td>TTC25</td>
<td>ENST00000377540</td>
<td>-</td>
</tr>
<tr>
<td> ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> 38</td>
<td>17</td>
<td>ENSG00000204815</td>
<td>TTC25</td>
<td>ENST00000593239</td>
<td>-</td>
</tr>
<tr>
<td> 38</td>
<td>17</td>
<td>ENSG00000204815</td>
<td>TTC25</td>
<td>ENST00000591658</td>
<td>-</td>
</tr>
<tr>
<td> 38</td>
<td>17</td>
<td>ENSG00000204815</td>
<td>TTC25</td>
<td>ENST00000377540</td>
<td>NM_031421</td>
</tr>
<tr>
<td> 38</td>
<td>17</td>
<td>ENSG00000204815</td>
<td>TTC25</td>
<td>ENST00000377540</td>
<td>NM_031421</td>
</tr>
<tr>
<td> 38</td>
<td>17</td>
<td>ENSG00000204815</td>
<td>TTC25</td>
<td>ENST00000585530</td>
<td>-</td>
</tr>
</tbody>
</table>
<h3 id="ptprq">PTPRQ</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>37</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000551042</td>
<td>-</td>
</tr>
<tr>
<td><strong>37</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000547376</td>
<td>-</td>
</tr>
<tr>
<td><strong>37</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000551573</td>
<td>-</td>
</tr>
<tr>
<td><strong>37</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000526956</td>
<td>-</td>
</tr>
<tr>
<td><strong>37</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000547485</td>
<td>-</td>
</tr>
<tr>
<td><strong>37</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000551624</td>
<td>-</td>
</tr>
<tr>
<td><strong>37</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000547881</td>
<td>-</td>
</tr>
<tr>
<td><strong>37</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000549355</td>
<td>-</td>
</tr>
<tr>
<td><strong>37</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000266688</td>
<td>-</td>
</tr>
<tr>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000623635</td>
<td>-</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000551042</td>
<td>-</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000547376</td>
<td>-</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000551573</td>
<td>-</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000614701</td>
<td>NM_001145026</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000547485</td>
<td>-</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000551624</td>
<td>-</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000547881</td>
<td>-</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000549355</td>
<td>-</td>
</tr>
<tr>
<td><strong>38</strong></td>
<td>12</td>
<td>ENSG00000139304</td>
<td>PTPRQ</td>
<td>ENST00000616559</td>
<td>XM_017019274</td>
</tr>
</tbody>
</table>
<h3 id="srd5a2">SRD5A2</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td>37</td>
<td>2</td>
<td>ENSG00000049319</td>
<td>SRD5A2</td>
<td>ENST00000405650</td>
<td>-</td>
</tr>
<tr>
<td>37</td>
<td>2</td>
<td>ENSG00000049319</td>
<td>SRD5A2</td>
<td>ENST00000233139</td>
<td>-</td>
</tr>
<tr>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td>38</td>
<td>2</td>
<td>ENSG00000277893</td>
<td>SRD5A2</td>
<td>ENST00000622030</td>
<td>NM_000348</td>
</tr>
</tbody>
</table>
<p>Here we can see that two transcripts has been merged to one and been given a refseq identifier. Most probable one of the old transcripts match the new one. We can not be sure if variants are missed here.</p>
<h3 id="pigy">PIGY</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td>37</td>
<td>4</td>
<td>ENSG00000255072</td>
<td>PIGY</td>
<td>ENST00000527353</td>
<td>-</td>
</tr>
<tr>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td>38</td>
<td>4</td>
<td>ENSG00000255072</td>
<td>PIGY</td>
<td>ENST00000527353</td>
<td>-</td>
</tr>
</tbody>
</table>
<p>Here the transcript is identical for both builds so we dear to guess that we do not miss any variants.</p>
<h3 id="fgf16">FGF16</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>X</td>
<td>ENSG00000196468</td>
<td>FGF16</td>
<td>ENST00000439435</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>HG1426_PATCH</td>
<td>ENSG00000268853</td>
<td>FGF16</td>
<td>ENST00000600602</td>
<td>-</td>
</tr>
<tr>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>X</td>
<td>ENSG00000196468</td>
<td>FGF16</td>
<td>ENST00000439435</td>
<td>NM_003868</td>
</tr>
</tbody>
</table>
<p> The refseq transcript exists in build 37 but is missing the refseq identifier. We would NOT miss any variants here.</p>
<h3 id="trac">TRAC</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> 37</td>
<td>14</td>
<td>ENSG00000229164</td>
<td>TRAC</td>
<td>ENST00000478163</td>
<td>-</td>
</tr>
<tr>
<td> ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> 38</td>
<td>14</td>
<td>ENSG00000277734</td>
<td>TRAC</td>
<td>ENST00000611116</td>
<td>-</td>
</tr>
</tbody>
</table>
<p> Transcripts have different ENS ids in both builds but are most probably the same. No refseq identifier for any build.</p>
<h3 id="padi6">PADI6</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <em>37</em></td>
<td>1</td>
<td>ENSG00000256049</td>
<td>PADI6</td>
<td>ENST00000434762</td>
<td>-</td>
</tr>
<tr>
<td> <em>37</em></td>
<td>1</td>
<td>ENSG00000256049</td>
<td>PADI6</td>
<td>ENST00000358481</td>
<td>-</td>
</tr>
<tr>
<td> ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <em>38</em></td>
<td>1</td>
<td>ENSG00000276747</td>
<td>PADI6</td>
<td>ENST00000619609</td>
<td>NM_207421</td>
</tr>
<tr>
<td> <em>38</em></td>
<td>CHR_HG2095_PATCH</td>
<td>ENSG00000280949</td>
<td>PADI6</td>
<td>ENST00000625380</td>
<td>NM_207421</td>
</tr>
</tbody>
</table>
<p> Unclear what happens here...</p>
<h3 id="gdf1">GDF1</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>19</td>
<td>ENSG00000130283</td>
<td>GDF1</td>
<td>ENST00000247005</td>
<td>-</td>
</tr>
<tr>
<td> ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>19</td>
<td>ENSG00000130283</td>
<td>GDF1</td>
<td>ENST00000247005</td>
<td>- </td>
</tr>
</tbody>
</table>
<p> Same transcripts, no refseq in both builds. No variants will be missed here</p>
<h3 id="tubb3">TUBB3</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>16</td>
<td>ENSG00000198211</td>
<td>TUBB3</td>
<td>ENST00000556922</td>
<td>-</td>
</tr>
<tr>
<td> ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000555810</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000554444</td>
<td>NM_001197181</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000556565</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000315491</td>
<td>NM_006086</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000553656</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000556536</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000554116</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000554927</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000557262</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000557490</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000555576</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000554336</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000555609</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000553967</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>16</td>
<td>ENSG00000258947</td>
<td>TUBB3</td>
<td>ENST00000625617</td>
<td>-</td>
</tr>
</tbody>
</table>
<p>This is very unclear, it goes from one to many transcripts between the builds. Hard to say if variants are missed here.</p>
<h3 id="ighg2">IGHG2</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>14</td>
<td>ENSG00000211893</td>
<td>IGHG2</td>
<td>ENST00000390545</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>HG1592_PATCH</td>
<td></td>
<td>ENSG00000270895</td>
<td>IGHG2</td>
<td>ENST00000603473</td>
<td>-</td>
</tr>
<tr>
<td> ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>CHR_HSCHR14_3_CTG1</td>
<td>ENSG00000274497</td>
<td>IGHG2</td>
<td>ENST00000621803</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>14</td>
<td>ENSG00000211893</td>
<td>IGHG2</td>
<td>ENST00000641095</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>14</td>
<td>ENSG00000211893</td>
<td>IGHG2</td>
<td>ENST00000390545</td>
<td>-</td>
</tr>
</tbody>
</table>
<p> No refseq in any build. Same transcripts in both builds except a new one in a patch. No variants would be missed here.</p>
<h3 id="ighm">IGHM</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>14</td>
<td>ENSG00000211899</td>
<td>IGHM</td>
<td>ENST00000390559</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>HG1592_PATCH</td>
<td>ENSG00000271541</td>
<td>IGHM</td>
<td>ENST00000605693</td>
<td>-</td>
</tr>
<tr>
<td> ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>CHR_HSCHR14_3_CTG1</td>
<td>ENSG00000282657</td>
<td>IGHM</td>
<td>ENST00000626472</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>14</td>
<td>ENSG00000211899</td>
<td>IGHM</td>
<td>ENST00000637539</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>14</td>
<td>ENSG00000211899</td>
<td>IGHM</td>
<td>ENST00000390559</td>
<td>- </td>
</tr>
</tbody>
</table>
<p> Same as above, no variants missing here</p>
<h3 id="igkc">IGKC</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>2</td>
<td>ENSG00000211592</td>
<td>IGKC</td>
<td>ENST00000390237</td>
<td>-</td>
</tr>
<tr>
<td>  ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>2</td>
<td>ENSG00000211592</td>
<td>IGKC</td>
<td>ENST00000390237</td>
<td>-</td>
</tr>
</tbody>
</table>
<p> Same in both builds. No variants would be missing here</p>
<h3 id="fcgr2c">FCGR2C</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000502411</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000496692</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000466542</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000465075</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000473530</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000473712</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000482226</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000467903</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000507374</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000508651</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000543859</td>
<td>-</td>
</tr>
<tr>
<td>  ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000502411</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000496692</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000466542</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000465075</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000473530</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000473712</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000482226</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000467903</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000507374</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000508651</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000611236</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>1</td>
<td>ENSG00000244682</td>
<td>FCGR2C</td>
<td>ENST00000543859</td>
<td>NR_047648</td>
</tr>
</tbody>
</table>
<p>Most probably no variants would be missing here. All transcripts seems to be correct between the builds.</p>
<h3 id="kmt2b">KMT2B</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>19</td>
<td>ENSG00000105663</td>
<td>KMT2B</td>
<td>ENST00000606995</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>19</td>
<td>ENSG00000105663</td>
<td>KMT2B</td>
<td>ENST00000607650</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>19</td>
<td>ENSG00000105663</td>
<td>KMT2B</td>
<td>ENST00000592092</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>19</td>
<td>ENSG00000105663</td>
<td>KMT2B</td>
<td>ENST00000585476</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>19</td>
<td>ENSG00000105663</td>
<td>KMT2B</td>
<td>ENST00000586308</td>
<td>-</td>
</tr>
<tr>
<td>  ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>19</td>
<td>ENSG00000272333</td>
<td>KMT2B</td>
<td>ENST00000420124</td>
<td>NM_014727</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>19</td>
<td>ENSG00000272333</td>
<td>KMT2B</td>
<td>ENST00000606995</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>19</td>
<td>ENSG00000272333</td>
<td>KMT2B</td>
<td>ENST00000592092</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>19</td>
<td>ENSG00000272333</td>
<td>KMT2B</td>
<td>ENST00000585476</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>19</td>
<td>ENSG00000272333</td>
<td>KMT2B</td>
<td>ENST00000586308</td>
<td>-</td>
</tr>
</tbody>
</table>
<p>Here the transcript <code>ENST00000607650</code> in build 37 have probably changed id to <code>ENST00000420124</code> and been given a refseq identifier.</p>
<h3 id="nefl">NEFL</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>8</td>
<td>ENSG00000104725</td>
<td>NEFL</td>
<td>ENST00000221169</td>
<td>-</td>
</tr>
<tr>
<td>  ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>8</td>
<td>ENSG00000277586</td>
<td>NEFL</td>
<td>ENST00000610854</td>
<td>NM_006158</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>8</td>
<td>ENSG00000277586</td>
<td>NEFL</td>
<td>ENST00000615973</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>8</td>
<td>ENSG00000277586</td>
<td>NEFL</td>
<td>ENST00000639464</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>8</td>
<td>ENSG00000277586</td>
<td>NEFL</td>
<td>ENST00000619417</td>
<td>-</td>
</tr>
</tbody>
</table>
<p>This is unclear, probably ENST00000221169 has been renamed to ENST00000610854 and been given a refseq identifier.</p>
<h3 id="nr2e3">NR2E3</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>15</td>
<td>ENSG00000031544</td>
<td>NR2E3</td>
<td>ENST00000561604</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>15</td>
<td>ENSG00000031544</td>
<td>NR2E3</td>
<td>ENST00000567496</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>15</td>
<td>ENSG00000031544</td>
<td>NR2E3</td>
<td>ENST00000562839</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>15</td>
<td>ENSG00000031544</td>
<td>NR2E3</td>
<td>ENST00000562925</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>15</td>
<td>ENSG00000031544</td>
<td>NR2E3</td>
<td>ENST00000563709</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>15</td>
<td>ENSG00000031544</td>
<td>NR2E3</td>
<td>ENST00000398840</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>15</td>
<td>ENSG00000031544</td>
<td>NR2E3</td>
<td>ENST00000326995</td>
<td>-</td>
</tr>
<tr>
<td>  ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>15</td>
<td>ENSG00000278570</td>
<td>NR2E3</td>
<td>ENST00000621736</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>15</td>
<td>ENSG00000278570</td>
<td>NR2E3</td>
<td>ENST00000617575</td>
<td>NM_014249</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>15</td>
<td>ENSG00000278570</td>
<td>NR2E3</td>
<td>ENST00000621098</td>
<td>NM_016346</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>15</td>
<td>ENSG00000278570</td>
<td>NR2E3</td>
<td>ENST00000563709</td>
<td>-</td>
</tr>
</tbody>
</table>
<p>This is unclear. We might miss variants here</p>
<h3 id="map3k14">MAP3K14</h3>
<table>
<thead>
<tr>
<th>Build</th>
<th>Chrom</th>
<th>EnsGeneID</th>
<th>Gene Name</th>
<th>EnsTransID</th>
<th>RefSeq mRNA</th>
</tr>
</thead>
<tbody>
<tr>
<td> <strong>37</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000587332</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000592267</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000586644</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000344686</td>
<td>-</td>
</tr>
<tr>
<td> <strong>37</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000376926</td>
<td>-</td>
</tr>
<tr>
<td>  ---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
<td>---</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000344686</td>
<td>NM_003954</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000640087</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000592267</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000586644</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000617331</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>17</td>
<td>ENSG00000006062</td>
<td>MAP3K14</td>
<td>ENST00000376926</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>CHR_HSCHR17_2_CTG5</td>
<td>ENSG00000282637</td>
<td>MAP3K14</td>
<td>ENST00000633437</td>
<td>-</td>
</tr>
<tr>
<td> <strong>38</strong></td>
<td>CHR_HSCHR17_2_CTG5</td>
<td>ENSG00000282637</td>
<td>MAP3K14</td>
<td>ENST00000634016</td>
<td>-</td>
</tr>
</tbody>
</table>
<p> Here one of the transcripts have been given a refseq identifier in build 38, that transcript exists in build 37 so no variants would be missed here. </p>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="variants.html" class="navigation navigation-prev " aria-label="Previous page: Variants">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
                <a href="panels.html" class="navigation navigation-next " aria-label="Next page: Gene Panels">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"Genes","level":"1.5.8","depth":2,"next":{"title":"Gene Panels","level":"1.5.9","depth":2,"path":"user-guide/panels.md","ref":"user-guide/panels.md","articles":[]},"previous":{"title":"Variants","level":"1.5.7","depth":2,"path":"user-guide/variants.md","ref":"user-guide/variants.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{},"plugins":[],"pluginsConfig":{"highlight":{},"search":{},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"fontsettings":{"theme":"white","family":"sans","size":2},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"user-guide/genes.md","mtime":"2018-06-13T07:46:06.281Z","type":"markdown"},"gitbook":{"version":"3.2.2","time":"2018-06-15T11:14:04.843Z"},"basePath":"..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

