# Scout
Analyze VCFs quicker and easier.

## Motivation
DNA sequencing is quickly entering the clinical setting. As such we need tools that cater to non-bioinformatics people. Scout is an attempt to provide an interface for bioinformaticians and clinicians alike to collaborate on the interpretation of variant calling results.


## Installation
Scout is not currently on *pip* but we intend on registering it once we work out more specifically the scope of the package.

For now, install as:

```bash
$ pip install https://github.com/Clinical-Genomics/scout/zipball/master
```

### Core dependencies

- MongoDB
- Python
- Probably some C-compiler(s)

**Recommended dependencies**

- Docker


## Goals
There are 3 main goals stated for the project.

  1. Visualize (annotated) VCFs
  2. Following (coding) best practices
  3. Be designed as a natural point of entry for customer

For each of these topics there are a number of more concrete objectives ranked as primary, secondary (if possibles), or tertiary (if we have time).

### 1. Visualize (annotated) VCFs
We don't need to be GEMINI, focus more on visual part, marking, commenting etc.


## Contributing
Further details to come at a later stage.
